src/cpp/benders/outer_loop/MasterUpdateBase.cpp

```
@@ -35,37 +25,42 @@ void MasterUpdateBase::CheckTau(double tau) {
-  const auto &obj = benders_->MasterObjectiveFunctionCoeffs();
-  const auto max_invest =
-      benders_->BestIterationWorkerMaster().get_max_invest();
-  lambda_max_ = 0;
-  for (const auto &[var_name, var_id] : benders_->MasterVariables()) {
-    lambda_max_ += obj[var_id] * max_invest.at(var_name);
-  }
-}
-void MasterUpdateBase::Update(const CRITERION &criterion) {
-    switch (criterion) {
-      case CRITERION::LOW:
-        lambda_max_ =
-            std::min(lambda_max_, benders_->GetBestIterationData().invest_cost);
-        break;
-      case CRITERION::HIGH:
-        lambda_min_ = lambda_;
-        break;
 
-      default:
-        return;
+bool MasterUpdateBase::Update(double lambda_min, double lambda_max) {
+  // if (is_criterion_high) {
+  //   lambda_min = lambda_;
+  // } else {
+  //   lambda_max =
+  //       std::min(lambda_max, benders_->GetBestIterationData().invest_cost);
+  // }
+  // WorkerMasterData workerMasterData = benders_->BestIterationWorkerMaster();
+  // double invest_cost = workerMasterData._invest_cost;
+  // double overall_cost = invest_cost + workerMasterData._operational_cost;
+  // if (outerLoopBiLevel_.Update_bilevel_data_if_feasible(
+  //         workerMasterData._cut_trace,
+  //         benders_
+  //             ->GetOuterLoopCriterionAtBestBenders() /*/!\ must be at best
+  //             it*/,
+  //         overall_cost)) {
+  //   lambda_max = std::min(lambda_max, invest_cost);
+  // } else {
+  //   lambda_min = lambda_;
+  // }
+
+  stop_update_ =
+      std::abs(lambda_max - lambda_min) < outer_loop_stopping_threshold_;
+  if (!stop_update_) {
+    lambda_ = dichotomy_weight_coeff_ * lambda_max +
+              (1 - dichotomy_weight_coeff_) * lambda_min;
+    UpdateConstraints();
   }
-  lambda_ = dichotomy_weight_coeff_ * lambda_max_ +
-            (1 - dichotomy_weight_coeff_) * lambda_min_;
-  UpdateConstraints();
+  return stop_update_;
 }
 
 void MasterUpdateBase::UpdateConstraints() {
```

Should we raise an exception if master is empty rather than doing nothing ?